protocol-units/dispute/src/RStarM.sol

```diff
@@ -0,0 +1,259 @@
+        Dispute storage dispute = disputes[disputeHash];
+        require(dispute.timestamp != 0, "Dispute not found");
+        require(dispute.state == DisputeState.SUBMITTED, "Dispute already resolved");
+        dispute.state = state;
+        emit DisputeResolved(disputeHash, state);
+    }
+
+    function submitProof(bytes32 blockHash, Receipt calldata receipt) external {
+        require(!verifiedProofs[blockHash].exists, "Proof already submitted for this block");
+        bool isValid = verifier.verify_integrity(receipt);
+        verifiedProofs[blockHash] = Proof(blockHash, isValid, true);
+        emit ProofSubmitted(blockHash, isValid);
+    }
+
+    function verify(bytes calldata seal, bytes32 imageId, bytes32 postStateDigest, bytes32 journalDigest)
+        public
+        view
+        returns (bool)
+    {
+        //require(verifiedProofs[blockHash].exists, "Proof not found");
+        Receipt memory receipt = Receipt(
+            seal,
+            ReceiptClaim(
+                imageId,
+                postStateDigest,
+                ExitCode(SystemExitCode.Halted, 0),
+                bytes32(0),
+                Output(journalDigest, bytes32(0)).digest()
+            )
+        );
+        return grothVerify(receipt);
+    }
+
+    function grothVerify(Receipt memory receipt) public view returns (bool) {
+        (uint256 claim0, uint256 claim1) = splitDigest(receipt.claim.digest());
+        Seal memory seal = abi.decode(receipt.seal, (Seal));
+        return this.verifyProof(seal.a, seal.b, seal.c, [CONTROL_ID_0, CONTROL_ID_1, claim0, claim1, BN254_CONTROL_ID]);
+    }
+
+    // The camel case here is not standard solidity practice. But we use it because its the implemntation of the interface.
+    function verify_integrity(Receipt memory receipt) public view returns (bool) {
+        (uint256 claim0, uint256 claim1) = splitDigest(receipt.claim.digest());
+        console2.log("claim0: ", claim0);
+        console2.log("claim1: ", claim1);
+        Seal memory seal = abi.decode(receipt.seal, (Seal));
+        console2.log("sealed");
+        bool is_verified = this.verifyProof(seal.a, seal.b, seal.c, [CONTROL_ID_0, CONTROL_ID_1, claim0, claim1, BN254_CONTROL_ID]);
+        console2.log("verified", is_verified);
+        return is_verified;
+    }
+
+    function submitOptimisticCommitment(bytes32 blockHash, bytes calldata stateCommitment) external {
+        require(validators[msg.sender].isRegistered, "Validator not registered");
+        OptimisticCommitment storage commitment = optimisticCommitments[blockHash];
+        if (commitment.validatorCount == 0) {
+            commitment.blockHash = blockHash;
+            commitment.stateCommitment = stateCommitment;
+            commitment.validatorCount = 1;
+        } else {
+            require(keccak256(commitment.stateCommitment) == keccak256(stateCommitment), "State commitment mismatch");
```

Comment:
This is actually a race right now instead of m of n. You would have that the subsequent committers have to  match the commitment of the first committer.

Instead, you'll want to track all commitments and check each time whether you have m of n agreeing on a state.

Even after the state is agreed upon, you'll want to commit to track other commitments for slashing.